pennylane/spin/lattice.py

```diff
@@ -190,7 +206,80 @@ def _generate_grid(self, neighbour_order):
             lattice_points.append(point)
             lattice_map.append(node_index)
 
-        return math.array(lattice_points), math.array(lattice_map)
+        return math.array(lattice_points), lattice_map
+
+    def _get_custom_edges(self, custom_edges, lattice_map):
+        """Generates the edges described in `custom_edges` for all unit cells.
+
+        Args:
+          custom_edges (Optional[list(list(tuples))]): Specifies the edges to be added in the lattice.
+              Default value is None, which adds the edges based on neighbour_order.
+              Each element in the list is for a separate edge, and can contain 1 or 2 tuples.
+              First tuple contains the index of the starting and ending vertex of the edge.
+              Second tuple is optional and contains the operator on that edge and coefficient
+              of that operator.
+          lattice_map (list[int]): A list to represent the node number for each lattice_point.
+
+        **Example**
+
+        Generates a square lattice with a single diagonal and assigns a different operation
+        to horizontal, vertical, and diagonal edges.
+        >>> n_cells = [3,3]
+        >>> vectors = [[1, 0], [0,1]]
+        >>> custom_edges = [
+                [(0, 1), ("XX", 0.1)],
+                [(0, 3), ("YY", 0.2)],
+                [(0, 4), ("XY", 0.3)],
+                ]
+        >>> lattice = Lattice(n_cells=n_cells, vectors=vectors, custom_edges=custom_edges)
+
+        """
+
+        if not all([len(edge) in (1, 2) for edge in custom_edges]):
+            raise TypeError(
+                """
+                The elements of custom_edges should be lists of length 1 or 2. Inside said lists should be a tuple that contains two lattice indices to represent the edge and, optionally, a tuple that represents the operation and coefficient for that edge.
+                Every tuple must contain two lattice indices to represent the edge
+                and can optionally include a list to represent the operation and coefficient for that edge.
+                """
+            )
+
+        edges = []
+        n_sl = len(self.positions)
+        nsites_axis = math.cumprod([n_sl, *self.n_cells[:0:-1]])[::-1]
+
+        for i, custom_edge in enumerate(custom_edges):
+            edge = custom_edge[0]
+
+            if edge[0] >= self.n_sites or edge[1] >= self.n_sites:
+                raise ValueError(
+                    f"The edge {edge} has vertices greater than n_sites, {self.n_sites}"
+                )
+
+            edge_operation = custom_edge[1] if len(custom_edge) == 2 else i
+            map_edge1 = lattice_map.index(edge[0])
+            map_edge2 = lattice_map.index(edge[1])
```

Comment:
Hmm I just realized `lattice_map` is a list not a dictionary. Generally i feel like maps are dictionaries 😅 so the naming is a little unconventional hence why I didn't catch this till now. I would also change the name of lattice_map to like lattice_list (? not a great name either, idk, but map is worse since it misleads).

`index()` is a function that runs in O(N) time, so it's not great to be running this nested in a for loop. Could we convert lattice_map from a list to a dictionary? E.g. from `[3, 5, 1, 7, 2, 4]` to a dictionary `{"3": 0, "5": 1..."4": 6}` and do `lattice_dict.get(edge[0])` instead? That would be O(1) time O(1) space. Although I'm not familiar if we would get duplicate keys, so you will have to watch out for that.

Reply:
There is a bit of a complication here, I intended to set it as a dictionary initially. It is about what to set as key and value, obvious way would be to set the index as key and coordinates as the values, this presents the problem of duplicated keys as we renumber the indices based on boundary conditions. Another way will be to set the lattice point as key and index as value, this won't work in the above case, when we intend to find index for given lattice coordinates.
I hope this answers your question.
lattice_list is not a good name to describe what this list does, so I named it lattice_map, I can probably call it lattice_indices, does that sound better?

Reply:
`index()` doesn't work for duplicate keys either though. When we have a list like [3, 5, 1, 3], and do list.index(3) I'm pretty sure it always returns the earliest instance of 3, which would be 0. Rather than call `lattice_map.index(edge[0])`, you can first turn `lattice_map` into a dictionary where any duplicate key found in the list is simply ignored, e.g. given `list [3, 5, 1, 3]`, the equivalent dictionary is `{"3": 0, "5": 1, "1": 3}`. Functionally, this makes `lattice_map.index(3)` functionally equivalent to `lattice_map.get("3")`. 

Naming wise lattice_indices is fine, but if we are able to turn it into a dictionary then lattice_map is accurate.

Reply:
If you needed to retain the information of the duplicate keys, then you can opt for a dictionary like this `{"3": [0, 3], "5": [1], "1": [3]}` (any similar implementation would work, such as with tuples). When you call lattice_map.get(), you can simply always get the 0th index of the array you get.